zerver/migrations/0382_revoke_invitations_from_deactivated_users.py

```diff
@@ -0,0 +1,77 @@
+from typing import List
+
+from django.db import migrations
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.utils.timezone import now as timezone_now
+
+
+def revoke_invitations(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    Confirmation = apps.get_model("confirmation", "Confirmation")
+    Confirmation.INVITATION = 2
+    Confirmation.MULTIUSE_INVITE = 6
+    PreregistrationUser = apps.get_model("zerver", "PreregistrationUser")
+    UserProfile = apps.get_model("zerver", "UserProfile")
+    MultiuseInvite = apps.get_model("zerver", "MultiuseInvite")
+
+    STATUS_ACTIVE = 1
+    STATUS_REVOKED = 2
+
+    def get_valid_invite_confirmations_generated_by_users(
+        user_ids: List[int],
+    ) -> List[int]:
+        prereg_user_ids = (
+            PreregistrationUser.objects.filter(referred_by_id__in=user_ids)
+            .exclude(status__in=[STATUS_ACTIVE, STATUS_REVOKED])
+            .values_list("id", flat=True)
+        )
+        confirmation_ids = list(
+            Confirmation.objects.filter(
+                type=Confirmation.INVITATION,
+                object_id__in=prereg_user_ids,
+                expiry_date__gte=timezone_now(),
+            ).values_list("id", flat=True)
+        )
+
+        multiuse_invite_ids = MultiuseInvite.objects.filter(
+            referred_by_id__in=user_ids
+        ).values_list("id", flat=True)
+        confirmation_ids += list(
+            Confirmation.objects.filter(
+                type=Confirmation.MULTIUSE_INVITE,
+                expiry_date__gte=timezone_now(),
+                object_id__in=multiuse_invite_ids,
+            ).values_list("id", flat=True)
+        )
+
+        return confirmation_ids
+
+    for realm_id in Realm.objects.values_list("id", flat=True):
+        deactivated_user_ids = UserProfile.objects.filter(
+            is_active=False, realm_id=realm_id
+        ).values_list("id", flat=True)
+        confirmation_ids = get_valid_invite_confirmations_generated_by_users(deactivated_user_ids)
+        Confirmation.objects.filter(id__in=confirmation_ids).update(expiry_date=timezone_now())
```

Comment:
It would be kinda nice to have logging for which confirmation objects were modified in this way, since this operation isn't easily reversed as is.  Do we have a good way to do that?

Reply:
We can just print the `confirmation_ids` I think? I added that

Reply:
Yep that works.